plutus-core/plutus-ir/test/GeneratorSpec.hs

```diff
@@ -0,0 +1,271 @@
+prop_genTypeCorrect :: Property
+prop_genTypeCorrect =
+  forAllDoc "ty,tm" genTypeAndTerm_ (const []) $ \ (ty, tm) -> typeCheckTerm tm ty
+
+-- | Test that when we generate a fully applied term we end up
+-- with a well-typed term.
+prop_genWellTypedFullyApplied :: Property
+prop_genWellTypedFullyApplied =
+  forAllDoc "ty, tm" genTypeAndTerm_ (shrinkTypedTerm mempty mempty) $ \ (ty, tm) ->
+  forAllDoc "ty', tm'" (genFullyApplied ty tm) (const []) $ \ (ty', tm') -> typeCheckTerm tm' ty'
+
+-- | Test that shrinking a well-typed term results in a well-typed term
+prop_shrinkTermSound :: Property
+prop_shrinkTermSound =
+  forAllShrinkBlind (pure False) (\ sh -> [ True | not sh ]) $ \ _ ->
+  forAllDoc "ty,tm"   genTypeAndTerm_ shrinkClosedTypedTerm $ \ (ty, tm) ->
+  let shrinks = shrinkClosedTypedTerm (ty, tm) in
+  -- While we generate well-typed terms we still need this check here for
+  -- shrinking counterexamples to *this* property. If we find a term whose
+  -- shrinks aren't well-typed we want to find smaller *well-typed* terms
+  -- whose shrinks aren't well typed.
+  typeCheckTerm tm ty ==>
+  not (null shrinks) ==>
+  checkNoCounterexamples [ (ty, tm, scopeCheckTyVars Map.empty (ty, tm))
+                         | (ty, tm) <- shrinks, not $ typeCheckTerm tm ty ]
+
+-- * Utility tests for debugging generators that behave weirdly
+
+-- | Test that `typeInstTerm` results in a well-typed instantiation.
+prop_typeInstTerm :: Property
+prop_typeInstTerm =
+  forAllDoc "ctx"    genCtx                      (const [])       $ \ ctx ->
+  forAllDoc "ty"     (genTypeWithCtx ctx $ Star) (shrinkType ctx) $ \ ty ->
+  forAllDoc "target" (genTypeWithCtx ctx $ Star) (shrinkType ctx) $ \ target ->
+  doTypeInstTermCheck ctx ty target
+  where
+    doTypeInstTermCheck :: Map TyName (Kind ())
+                        -> Type TyName DefaultUni ()
+                        -> Type TyName DefaultUni ()
+                        -> Property
+    doTypeInstTermCheck ctx ty target =
+      case [ (n, insts)
+           | n <- [0..arity ty+3]
+           , Just insts <- [typeInstTerm ctx n target ty]
+           , not $ checkInst ctx x ty insts target
+           ] of
+        []  -> property True
+        bad -> ceDoc (prettyPirReadable bad) False
+      where
+        x = Name "x" (toEnum 0)
+        arity (TyForall _ _ _ a) = arity a
+        arity (TyFun _ _ b)      = 1 + arity b
+        arity _                  = 0
+
+        checkInst ctx x ty insts target = typeCheckTermInContext ctx tmCtx tm target
+          where
+            (tmCtx, tm) = go (toEnum 1) (Map.singleton x ty) (Var () x) insts
+            go _ tmCtx tm [] = (tmCtx, tm)
+            go i tmCtx tm (InstApp ty : insts) = go i tmCtx (TyInst () tm ty) insts
+            go i tmCtx tm (InstArg ty : insts) = go (succ i) (Map.insert y ty tmCtx)
+                                                             (Apply () tm (Var () y)) insts
+              where y = Name "y" i
```

Comment:
I think this is a good idea. It might requrie a bit of an overhaul of the generators to incorporate the `Quote` stuff "from the bottom up" - but it might help readability and reduce code duplication.